### PR TITLE
Fix a bug in the register allocator and a few other bits.

### PR DESCRIPTION
**Pull request by vext01**
See individual commits.

This will clash with the other PR, so I'll probably have to resolve some conflicts later.

### CODE REVIEW

File: ykcompile/src/lib.rs [hunk 1]
```diff
@@ -36,44 +36,52 @@ pub struct TraceCompiler {
...
+            if self.assigned_regs.contains_key(&l) {
+                self.assigned_regs[&l]
+            } else {
+                let reg = self.available_regs.pop().expect(
+                    "Can't allocate more than 7 locals
+                    yet!",
+                );
+                self.assigned_regs.insert(l, reg);
+                reg
+            }
```
Comment (bjorn3) [inline ykcompile/src/lib.rs]:
```suggestion
            let available_regs = &mut self.available_regs;
            *self.assigned_regs.entry(l).or_insert_with(|| {
                available_regs.pop().expect("Can't allocate more than 7 locals yet!");
            })
```

Comment (vext01) [reply ykcompile/src/lib.rs]: I tried something similar to this and it whined about mutable borrows inside the `or_insert_with` closure. Let's give it a shot.

Comment (vext01) [reply ykcompile/src/lib.rs]: Worked with a minor tweak :)

I think the trick is to do `let available_regs = &mut self.available_regs`.

I had repeated `self.available_regs` inside the closure, which upset the borrow checker.

Thanks!

Comment (bjorn3) [reply ykcompile/src/lib.rs]: > I think the trick is to do let available_regs = &mut self.available_regs.

See my code :) A closure currently captures a whole variable (`self`), not just a part of it (`self.available_regs`). There is an accepted RFC to only capture the used part, but the implementation is currently on hold.

Comment (bjorn3) [reply ykcompile/src/lib.rs]: https://github.com/rust-lang/rfcs/pull/2229

File: ykcompile/src/lib.rs [hunk 2]
```diff
@@ -45,7 +45,10 @@ impl TraceCompiler {
         // locals. We can do better than this by using StorageLive/StorageDead from the MIR to free
         // up registers again, and allocate additional locals on the stack. Though, ultimately we
         // probably want to implement a proper register allocator, e.g. linear scan.
+
         if l == 0 {
+            // In SIR, `Local` zero is the (implicit) return value, so it makes sense to allocate it
+            // in return register of the underlying X86_64 calling convention.
```
Comment (ptersilie) [inline ykcompile/src/lib.rs]: `in the return register` (missing `the`)

**Comment by ptersilie** [comment]
One typo, rest looks good.

**Comment by vext01** [comment]
Fixed typo, but we this will need re-sync after #55

**Comment by ptersilie** [comment]
Looks good, squash and rebase once #55 is done.

**Comment by vext01** [comment]
That was quite involved to rebase!

We can't use the entry API after all, as there's no way to get the `OutOfRegisters` error out of the `or_insert_with` closure.

**Pull request merged**